### PR TITLE
✨Elliptic Curve Cryptography

### PR DESCRIPTION
**Pull request by raxhvl**
Foundations of ECC. Preview the markdown to render media.

### CODE REVIEW

**Review by taxmeifyoucan — COMMENTED** [review]
Thank you so much for the contribution @raxhvl! Incredible job on this one, it's comprehensive and understandable. I would maybe do some tweaks to the intro but it's great writing! Keep them coming :)

**Comment by raxhvl** [comment]
> Thank you so much for the contribution @raxhvl! Incredible job on this one, it's comprehensive and understandable. I would maybe do some tweaks to the intro but it's great writing! Keep them coming :)

Thank your for the kind words! Now the real challenge is writing an intro that lives up to the rest (wink wink).

File: docs/wiki/Cryptography/elliptic-curve-cryptography.md [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+Alice intends to sign the message **"Send $1 million"**, by following the steps:
+
+1. Compute the cryptographic hash **$m$**.
+
+```python
+sage: m = hash("Send $1 million")
+-7930066429007744594
+```
+
+2. For every signature, a random **ephemeral key pair [$eK$, $eP$]** is generated to mitigate an [attack](https://youtu.be/DUGGJpn2_zY?si=4FZ3ZlQZTG9-eah9&t=2117) exposing her **secret key 🔑**.
```
[the remark this answers is not in the document]
Comment (raxhvl) [reply docs/wiki/Cryptography/elliptic-curve-cryptography.md:177]: Intersting.. It works fine without the space for me.

Screenshot from 2024-02-27 12-04-21

@taxmeifyoucan could you check locally?

File: docs/wiki/Cryptography/elliptic-curve-cryptography.md [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+In mathematics, a **field** is a set $F$, containing at least two elements, which is closed under two binary operations usually referred to as **addition** ($+$), and **multiplication**($\times$). A set is closed under an operation when the result of the operation is also a member of the set.
+
+The set of real numbers $\mathbb{R}$ is a familiar example of field, since arithmetic addition of two real numbers is closed.
+
+$$
+ 3 \in \mathbb{R},  5 \in \mathbb{R} \\
+ 3 + 5 = 8 \in \mathbb{R}
+$$
+
+Elliptic curves are interesting because the points form a field, the result of "addition" of two points remains on the curve. This geometric addition, distinct from arithmetic counterparts, involves drawing a line through chosen points (**P** and **Q**) and reflecting the resulting curve intersection(**R'**) across the x-axis to yield their sum (**R**).
```
[the remark this answers is not in the document]
Comment (raxhvl) [reply docs/wiki/Cryptography/elliptic-curve-cryptography.md:48]: Nice catch! Thanks for carefully reviewing. I have:
1. Corrected the highlighted sentence.
2. Added a subsection for group.

File: docs/wiki/Cryptography/elliptic-curve-cryptography.md [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+The goal of this exercise is device a method for Alice to create a **secret key 🔑** known only to her. This key will be crucial for her to prove her identity and ensure the message's authenticity to Bob.
+
+Mathematics, as always, comes to the rescue. Through ingenious use of **Elliptic Curves**, let's explore how Alice can generate the **secret key 🔑**.
+
+# Elliptic curves
+
+An elliptic curve is a curve **described by the equation**:
+
+$$
+y^2 = x^3 + ax+b
```
[the remark this answers is not in the document]
Comment (raxhvl) [reply docs/wiki/Cryptography/elliptic-curve-cryptography.md:23]: Done! Let me know if it looks good.

File: docs/wiki/Cryptography/elliptic-curve-cryptography.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (SiddharthV1) [reply docs/wiki/Cryptography/elliptic-curve-cryptography.md:190]: I checked locally , it renders fine without the space

[the remark this answers is not in the document]
Comment (raxhvl) [reply docs/wiki/Cryptography/elliptic-curve-cryptography.md:190]: Thanks @SiddharthV1

File: docs/wiki/Cryptography/elliptic-curve-cryptography.md [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+Ethereum uses a standard curve known as [secp256k1](http://www.secg.org/sec2-v2.pdf) with parameters $a=0$, and $b=7$; which is the curve:
+$$y^2=x^3+7$$
+
+<img src="images/elliptic-curves/secp256k1.png" width="500"/>
+
+## Groups and Fields
+
+### Group
+In mathematics, a **GROUP** is a set $G$, containing at least two elements, which is closed under a binary operation usually referred to as **addition** ($+$). A set is closed under an operation when the result of the operation is also a member of the set.
```
Comment (SiddharthV1) [inline docs/wiki/Cryptography/elliptic-curve-cryptography.md:47]: Shouldn't something like this be also included in group definition?   ... Groups have an associative operation and an identity element

File: docs/wiki/Cryptography/elliptic-curve-cryptography.md [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+$$
+ 3 \in \mathbb{R},  5 \in \mathbb{R} \\
+ 3 + 5 = 8 \in \mathbb{R}
+$$
+
+## Field
+Similarly, a **FIELD** is a set $F$, containing at least two elements, which is closed under two binary operations usually referred to as **addition** ($+$), and **multiplication**($\times$). 
+
+In other words, A **FIELD** is a **GROUP** under both addition and multiplication.
```
Comment (SiddharthV1) [inline docs/wiki/Cryptography/elliptic-curve-cryptography.md:59]: Would it be good to include something like this : A field is an commutative ring containing  multiplicative inverses for non zero elements . Additionally, I think I found a good link giving all three definitions groups , rings and fields on a page : https://www.brainkart.com/article/Groups,-Rings,-and-Fields_8402/

File: docs/wiki/Cryptography/elliptic-curve-cryptography.md (hunk 4 above)
Comment (raxhvl) [reply docs/wiki/Cryptography/elliptic-curve-cryptography.md:47]: Ideally it should be. My intent is to make as simple as possible for everyone, regardless of their math background. This definition assumes the reader knows about what inverses while the knowledge of inverse and associative is not applied in the context of ECDSA. Does that sound okay?

File: docs/wiki/Cryptography/elliptic-curve-cryptography.md (hunk 5 above)
Comment (raxhvl) [reply docs/wiki/Cryptography/elliptic-curve-cryptography.md:59]: Great resource. The diagram towards the bottom is especially usefull. 

Do you have a reference which cover some examples too maybe like this one: https://www.rareskills.io/post/group-theory-and-coding ?

Comment (SiddharthV1) [reply docs/wiki/Cryptography/elliptic-curve-cryptography.md:59]: Yes that's a great idea! let me poke around , if I find something worthwhile I'll post it in this comments thread

**Comment by raxhvl** [comment]
hey @taxmeifyoucan I have updated the article:
1. Added a super short intro to public key cryptography.
2. Introduced ECDSA in the context of Ethereum.
3. Added a link to a post that discusses how raw transactions are signed using ECDSA.

**Comment by taxmeifyoucan** [comment]
Thanks a lot for the content and updates @raxhvl! It looks good to me, I will add it to the next wiki release

**Pull request merged**